Set settings section as selected if user presses the sidebar button

LittleEndu opened the pull request:
Closes #9196

![settings](https://user-images.githubusercontent.com/11320483/100555208-8b69f580-32a2-11eb-8098-479e3539b8b7.gif)

~~I assume the section we scroll to has to reach the search box to be considered selected but currently it will never reach that (at least not high enough to always satisfy ``<= 0``)~~

~~Considering the margin when deciding what section is selected sounds most reasonable, but maybe this should be even more than ``51`` because users would consider the section reached well before.~~

LittleEndu committed:
Consider section margin when highlighting

bdach commented:
I don't think this is a correct solution. This is actually an issue with the comparison-to-zero which suffers from floating-point precision issues. I'm claiming that based on the following log excerpt I acquired when I logged what the `diff` function returns for each section when navigating between them:

```
[runtime] 2020-11-29 18:08:01 [debug]: MouseDownEvent(Left) handled by Sidebar+SidebarScrollContainer.
[runtime] 2020-11-29 18:08:01 [debug]: ClickEvent(Left) handled by SidebarButton.
[runtime] 2020-11-29 18:08:01 [debug]: MouseClick handled by SidebarButton.
[runtime] 2020-11-29 18:08:01 [verbose]: diffs: 120.26445, 476.26434, 1325.2622, 2577.262, 3420.262, 3782.2617, 4199.2617, 4386.2617, 4960.2617
[runtime] 2020-11-29 18:08:01 [verbose]: diffs: 58.10704, 414.10693, 1263.1049, 2515.1045, 3358.1045, 3720.1042, 4137.1045, 4324.1045, 4898.104
[runtime] 2020-11-29 18:08:01 [verbose]: diffs: 53.97528, 409.97516, 1258.973, 2510.973, 3353.973, 3715.9727, 4132.9727, 4319.9727, 4893.9727
(...)
[runtime] 2020-11-29 18:08:03 [verbose]: diffs: -355.9988, 0.0010986328, 848.999, 2100.9988, 2943.9988, 3305.9988, 3722.9988, 3909.9988, 4483.998
[runtime] 2020-11-29 18:08:03 [verbose]: diffs: -355.9988, 0.0010681152, 848.999, 2100.9988, 2943.9988, 3305.9988, 3722.9988, 3909.9988, 4483.9985
[runtime] 2020-11-29 18:08:03 [verbose]: diffs: -355.99988, 3.0517578E-05, 848.9979, 2100.9976, 2943.9978, 3305.9976, 3722.9978, 3909.9978, 4483.9976
[runtime] 2020-11-29 18:08:12 [debug]: MouseDownEvent(Left) handled by Sidebar+SidebarScrollContainer.
[runtime] 2020-11-29 18:08:13 [debug]: ClickEvent(Left) handled by SidebarButton.
[runtime] 2020-11-29 18:08:13 [debug]: MouseClick handled by SidebarButton.
[runtime] 2020-11-29 18:08:13 [verbose]: diffs: -355.99988, 6.1035156E-05, 848.9994, 2100.999, 2943.9985, 3305.9985, 3722.9985, 3909.999, 4483.999
[runtime] 2020-11-29 18:08:13 [debug]: MouseDownEvent(Left) handled by Sidebar+SidebarScrollContainer.
[runtime] 2020-11-29 18:08:14 [debug]: ClickEvent(Left) handled by SidebarButton.
[runtime] 2020-11-29 18:08:14 [debug]: MouseClick handled by SidebarButton.
[runtime] 2020-11-29 18:08:14 [verbose]: diffs: -355.99997, -6.1035156E-05, 848.9992, 2100.999, 2943.9988, 3305.9983, 3722.998, 3909.9985, 4483.9985
[runtime] 2020-11-29 18:08:14 [debug]: MouseDownEvent(Left) handled by Sidebar+SidebarScrollContainer.
[runtime] 2020-11-29 18:08:14 [debug]: ClickEvent(Left) handled by SidebarButton.
[runtime] 2020-11-29 18:08:14 [debug]: MouseClick handled by SidebarButton.
[runtime] 2020-11-29 18:08:14 [verbose]: diffs: -355.99988, 6.1035156E-05, 848.9994, 2100.999, 2943.9985, 3305.9985, 3722.9985, 3909.999, 4483.999
[runtime] 2020-11-29 18:08:17 [debug]: MouseDownEvent(Left) handled by Sidebar+SidebarScrollContainer.
[runtime] 2020-11-29 18:08:17 [debug]: ClickEvent(Left) handled by SidebarButton.
[runtime] 2020-11-29 18:08:17 [debug]: MouseClick handled by SidebarButton.
[runtime] 2020-11-29 18:08:17 [verbose]: diffs: -355.99997, -6.1035156E-05, 848.9992, 2100.999, 2943.9988, 3305.9983, 3722.998, 3909.9985, 4483.9985
[runtime] 2020-11-29 18:08:17 [debug]: MouseDownEvent(Left) handled by Sidebar+SidebarScrollContainer.
[runtime] 2020-11-29 18:08:17 [debug]: ClickEvent(Left) handled by SidebarButton.
[runtime] 2020-11-29 18:08:17 [debug]: MouseClick handled by SidebarButton.
```

Note that at the first click, the diff value never satisfies `<= 0`, and then begins to oscillate on subsequent clicks. Each button actually behaves slightly differently (some of them will reach 0, but only after an inordinately long wait).

LittleEndu commented:
I get that it's a floating-point precision issue, but I also think the section should be considered reached well before the diff satisfies ``<= 0``. ``Margin.Top`` (or ``51`` which it equals to) starts to sound more and more like an arbitrary number in my head, maybe something more like 1/4th the overlay height would be a better number to use.

bdach commented:
all of these sound more arbitrary to me than 0 if I'm honest.

LittleEndu commented:
I'm suggesting something way bigger than 0 because I think users would consider the "Skin" section reached on this screenshot and would find it weird that the button isn't highlighted.
![osu_2020-11-29_22-14-42](https://user-images.githubusercontent.com/11320483/100552517-58b70180-3290-11eb-85a5-90b762b62029.png)

I don't think something like some % of container height is arbitrary for this.

bdach commented:
then why isn't "input" reached? if we're going that far, then surely all sections which have a part of them visible on the screen should be highlighted, no?

(side note, just noticed that the positioning of the "seasonal backgrounds" options makes zero sense. it's in audio settings.)

LittleEndu commented:
> then surely all sections which have a part of them visible on the screen should be highlighted

#9594 ?

bdach commented:
well if that was rejected, then i'd say it should either be reaching the search box as is currently, or an entirely different measure of determining selection should be used (i.e. the section which has most of it visible on screen at the current moment is selected). that's way less arbitrary than saying "the section is selected when you cross the top 25% of the screen", because that arbitrary percentage has no visible reference point.

LittleEndu committed:
Refactor code into UserTrackingScrollContainer

Always highlight selected section when user clicked on it

Fix floating-point imprecision

Issue in comment seems to be fixed

LittleEndu committed:
Merge branch 'master' of https://github.com/ppy/osu into settings-sidebar-button

bdach commented on osu.Game/Overlays/UserProfileOverlay.cs:
```diff
@@ -202,7 +202,7 @@ public ProfileSectionsContainer()
             }
 
-            protected override OsuScrollContainer CreateScrollContainer() => new OverlayScrollContainer();
+            protected override UserTrackingScrollContainer CreateScrollContainer() => new OverlayScrollContainer();
```
this change seems unnecessary

LittleEndu replied:
Can change that and cast in ``SectionsContainer`` ctor instead if that's preferred
```cs
scrollContainer = (UserTrackingScrollContainer)CreateScrollContainer().With(s =>
                {
                    s.RelativeSizeAxes = Axes.Both;
                    s.Masking = true;
                    s.ScrollbarVisible = false;
                    s.Child = scrollContentContainer = CreateScrollContentContainer();
                })
```

bdach replied:
no, that's much worse - never mind, thought that the `UserProfileOverlay` change was inheriting from a different type

LittleEndu replied:
hmmm, but wouldn't something like this make more sense:
```cs
                else if (!(scrollContainer is UserTrackingScrollContainer user) || user.UserScrolling)
                {
                    SelectedSection.Value = Children.TakeWhile(section => diff(section) <= 1).LastOrDefault()
                                            ?? Children.FirstOrDefault();
                }
```

bdach replied:
I think that's unnecessary complication.

bdach commented on osu.Game/Screens/Select/BeatmapCarousel.cs:
```diff
@@ -889,28 +889,10 @@ protected override void PerformSelection()
-            /// </summary>
-            public bool UserScrolling { get; private set; }
-
-            // ReSharper disable once OptionalParameterHierarchyMismatch 2020.3 EAP4 bug. (https://youtrack.jetbrains.com/issue/RSRP-481535?p=RIDER-51910)
```
why was this removed? have you checked if this is resolved in rider EAP?

LittleEndu replied:
yes it was. and it didn't trigger after removing so I guess the tracker is correct

peppy commented:
If you're going to go in this direction and remove the animation during scroll (i think it feels okay without it so i'm neutral), why not fix it in a permanent way - match on the last clicked section rather than Y coordinates.

Your change doesn't fix all cases, see this scenario:

![2020-11-30 18 32 11](https://user-images.githubusercontent.com/191335/100592446-7640a580-333a-11eb-9433-9650a82170e4.gif)

LittleEndu commented:
I think I'll try to implement what bdach suggested, as in have the section which takes up the most of visible area consider as selected, and see how that looks. Something like that should look better, fix the issue, keep the animation, and would not introduce a new class.
